### PR TITLE
Revise README.md

### PR DESCRIPTION
**Pull request by smileyung**
python train.py ^GSPC 10 1000

### CODE REVIEW

**Comment by xtr33me** [comment]
@mdcw9 Unfortunately I haven't had much time to look more at the project since the other night but I should start back on it after my current RCNN training is complete.  That said, I would assume if you were to plot a running total of the outputs from the model via matplotlib, you should get exactly the graph you are looking for.  I personally would also plot the actual stock itself so you have a nice reference on the same graph.

I myself am going to be training against some crypto data (probably Bitcoin) since it seems to be a catalyst for the altcoins and see if I can use the same model to then predict others.  This is all new to me as well, so will be figuring it out along the way.  Hopefully with others working on this we can figure something out :D

**Comment by mdcw9** [comment]
@xtr33me No worries. I have been looking at it this evening. 

Perhaps I am being an idiot, but if you look at the results Siraj obtains, in addition to the code he actually uses during the video, it looks like the RL algorithm is making a guess at the value of the stock itself, and compares this guess with the actual stock value. If the guess is either larger or smaller than the actual stock value, the loss is treated in the same way. A gradient of decent is then used to minimise this loss.

The RL algorithm reported here, is very different. In this case, we are not interested in predicting the exact stock price, but only at what times to buy and sell. Essentially, only the trend matters in this case.

I don't know how it is possible that Siraj used this code.

I don't know how to make sense out of what plot you would make, all it shows you is when it buys and sells and holds per function time. It will not tell you what it thinks the stock price will be in the future.

Or am i being very dumb here?

**Comment by smileyung** [comment]
@mdcw9 
I think it is an interesting issue about trading strategy,
if you want to know when should I buy this stock, and when should I sell this stock, 
maybe I can use this model for the trading with good benefits.
(trading problem)

If you want to predict the exact stock price, the RNN or LSTM model will be suitable for time-series prediction.
(time-series prediction problem)

**Comment by mdcw9** [comment]
@smileyung I agree with you. However, if you look at the plot in the youtube video, you can see that the Tensorflow prediction is clearly independent of the real data. The tensorflow prediction is not just saying 'Buy here'or 'Sell here', it is providing a value for the stock price. My problem is, is that this code cannot do that, and thus cannot be the code Siraj used. Unless I am missing something of course.

**Comment by smileyung** [comment]
@mdcw9 can you share the youtube web of Siraj? 
I am also interested in it.

**Comment by mdcw9** [comment]
@xtr33me I found the exact same thing last night! Only on a different platform. Check this out: https://hub.packtpub.com/develop-stock-price-predictive-model-using-reinforcement-learning-tensorflow/

It uses Yahoo finance however, which i think no longer works? I was struggling getting even the first calls of the stocks to work. I tried looking for a google_finance module, but the only one i could find gives the current stock values only, not years prior.

I have ran out of time this week to get any further, but will come back to it on Monday. What my plan is, is to try and add stocks from hard disk, like this code does. I will then try and use the algorithm from this book.

Have you had any luck replicating what Siraj's graph looks like?

**Comment by smileyung** [comment]
default

I try to do the experiment on the return with epochs, 
find that the return doesn't converge,
what should I do?

**Comment by xtr33me** [comment]
@smileyung That's interesting...not quite sure on why you might be seeing this. I currently was intrigued by why Siraj's video overall was referencing different logic than what was shared, so I was working on that currently here: https://github.com/xtr33me/Reinforcement_Learning_For_Predictive_Analytics/blob/master/PredictiveAnalyticsCrypto.ipynb

I wanted to have something representative to the one in the video so I might better understand some possible gaps I am missing in the overall understanding.  Next I want to see if I can take into consideration Volume in that I might be able to mimic VPA to more accurately be able to detect manipulation.  I'm at work now, but I will take a look tonight at your git to see if I can better understand what might be happening above.

**Comment by xtr33me** [comment]
@smileyung Sorry but I took a look at your fork and didn't notice anything different.  What was the dataset you are using and how many data points do you have?

**Comment by smileyung** [comment]
@xtr33me I use the same data set 
train: ^GSPC
test: ^GSPC_2011

and after training, I evaluate the performance of model with different epochs (model_ep10, model_ep20, model_ep30, ..., model_ep1000)
but the result seems not well.

**Comment by smileyung** [comment]
@xtr33me 
I have 3 questions.

1st, in the Reinforcement_Learning_for_Stock_Prediction/train.py Line 37
reward = max(data[t] - bought_price, 0),
why we need to have max? can't we define the reward as data[t] - bought_price?

2nd, can't I have sell as first action? (Short!)

3rd, I didn't find "close Position" in the program.
(["close Position"](https://www.investopedia.com/terms/c/closeposition.asp))

**Comment by xtr33me** [comment]
Hey @smileyung ! 

To put some thoughts around your questions, please look below:
1st: I agree with you to a point....the only reason I can think that the original dev used max was because you can pass a default if an entry is empty...in this case 0 -- my assumption mind you :)

2nd: You by all means can short first :D , this will just require some rethinking of the original logic. In the readme I stated that in the future I would probably look at finding a more statistical way of buying in.  That said, you also could write logic that allows you to short, but the logic that would need to be modified in some way is in train.py line 35. Essentially since agent.inventory had nothing in it, the model wouldn't allow you to sell anything that you didn't already have. This was why for me the easiest approach to get something working was to populate it with a buy :)  However I knew this wasn't the best course of action for the long term.

3rd: In this project, "sell" is the close. ["Sell Position](https://www.investopedia.com/terms/s/sell.asp)

**Comment by xtr33me** [comment]
@smileyung 
I think this is an absolutely great idea.  I will start on getting something put in this week probably after work. One other area that I would love to get your thoughts around is possibly integrating the ability to take into consideration Volume.  Some years back I read a book by Anna Coulling on VPA "Volume Price Analysis" and it was extremely eye opening in the sense that you were better able to predict market manipulation.  Since the Yahoo dataset has Volume, I'm thinking that this might be an interesting feature to put in as well.

**Comment by xtr33me** [comment]
So I took a look last night and I am going to have to look a bit more into using Tensorboard with Keras as I have yet to do this as of yet.

**Comment by mdcw9** [comment]
I haven't had any time to recently to look at this, but what I want to do is set a rule where the algorithm can only trade once per week. This means that the best time to buy and sell would actually be useful for a user. 

When you use the evaluation data, is there a way to get predictions for the future, beyond the evaluation data? All we need is a decision of what to do 1 week after the evaluation set has stopped.

**Comment by mdcw9** [comment]
Another thing to do would be: instead of having three options of what to do, have a random number generator fixed between 1 and 0, with 3 decimal places. The stock price is normalised and the algorithm must predict the price. Positive reward is given for a smaller difference between the actual normalised stock price, and the predicted price.

This may well not work however.

**Comment by xtr33me** [comment]
Hey @smileyung 

The original dev on this project stated this was Q-learning.  There is a pretty good explanation here on the differences: https://bit.ly/2K66Tje  However I'm not going to lie to you...they are very similar and I am going to have to find some other resources to get a better grip on the major differences between the two. It seems to be around on/off policy, but I still have to get my head wrapped around it a bit more.  If you prefer a sample video that had a pretty decent explanation, check this out: https://www.coursera.org/lecture/practical-rl/on-policy-vs-off-policy-experience-replay-YdFw3

Regarding your other question, Sigmoid is used because it allows for us to put a restrictive bounds on our final value that will be between 0 and 1.Where if we did not put that restriction, there is a chance that we may step out of that range which would make it more difficult to learn.  When we use sigmoid, it is often used when our final output is representative of a yes or no, true or false, black or white type answer. I hope that helps some.

**Comment by smileyung** [comment]
@xtr33me 

thanks a lot, I have another idea about github issue:
Can we have the github issue on it to discuss the problem or some idea?

image

**Comment by smileyung** [comment]
image

**Comment by xtr33me** [comment]
Thanks @smileyung ! Was much easier than setting up a project 👍
